shared/libraries/native_streaming_protocol/src/base_session_handler.cpp

```diff
@@ -255,4 +264,95 @@ std::string BaseSessionHandler::getStringFromData(const void* source, size_t str
+            return createReadHeaderTask();
+        }
+
+        // Get packet buffer header from received buffer
+        packetBufferHeader = static_cast<GenericPacketHeader*>(std::malloc(headerSize));
+        copyData(packetBufferHeader, data, headerSize, bytesDone, size);
+        LOG_T("Received packet buffer header: header size {}, payload size {}",
+              packetBufferHeader->size, packetBufferHeader->payloadSize);
+        bytesDone += headerSize;
+
+        // Get packet buffer payload from received buffer
+        if (packetBufferHeader->payloadSize > 0)
+        {
+            packetBufferPayload = std::malloc(packetBufferHeader->payloadSize);
+            copyData(packetBufferPayload, data, packetBufferHeader->payloadSize, bytesDone, size);
+        }
+        else
+        {
+            packetBufferPayload = nullptr;
+        }
+    }
+    catch (const DaqException& e)
+    {
+        LOG_E("Protocol error: {}", e.what());
+        errorHandler(std::string("Protocol error - readPacket - ") + e.what(), session);
+        return createReadStopTask();
+    }
+
+    auto recvPacketBuffer =
+        std::make_shared<PacketBuffer>(packetBufferHeader,
+                                       packetBufferPayload,
+                                       [packetBufferHeader, packetBufferPayload]()
+                                       {
+                                           std::free(packetBufferHeader);
+                                           if (packetBufferPayload != nullptr)
+                                               std::free(packetBufferPayload);
+                                       });
+
+    packetBufferReceivedHandler(recvPacketBuffer);
+
+    return createReadHeaderTask();
+}
+
+void BaseSessionHandler::sendPacketBuffer(const PacketBufferPtr& packetBuffer)
+{
+    std::vector<WriteTask> tasks;
+
+    // create write task for packet buffer header
+    boost::asio::const_buffer packetBufferHeader(packetBuffer->packetHeader,
+                                                 packetBuffer->packetHeader->size);
+    WriteHandler packetBufferHeaderHandler = [packetBuffer]() {};
+    tasks.push_back(WriteTask(packetBufferHeader, packetBufferHeaderHandler));
+
+    if (packetBuffer->packetHeader->payloadSize > 0)
+    {
+        // create write task for packet buffer payload
+        boost::asio::const_buffer packetBufferPayload(packetBuffer->payload,
+                                                      packetBuffer->packetHeader->payloadSize);
+        WriteHandler packetBufferPayloadHandler = [packetBuffer]() {};
+        tasks.push_back(WriteTask(packetBufferPayload, packetBufferPayloadHandler));
+    }
+
+    // create write task for transport header
+    size_t payloadSize = calculatePayloadSize(tasks);
+    auto writeHeaderTask = createWriteHeaderTask(PayloadType::PAYLOAD_TYPE_STREAMING_PACKET, payloadSize);
+    tasks.insert(tasks.begin(), writeHeaderTask);
```

Comment:
I have seen this insert at the beginning a lot. I think this triggers for each existing WriteTask to be moved into the next element in the vector. If it turns that this is a performance issue later on, it might be wise to think about other data structure (i.e. linked list), or we can try to reserve a vector site before hand. Just something to think about.

Reply:
Agreed, this is a potential area for optimizations.